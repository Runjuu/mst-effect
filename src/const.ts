export const Identity: unique symbol = Symbol.for('mst-effect-identity')

export const HANDLE_MST_EFFECT_ACTIONS = 'HANDLE_MST_EFFECT_ACTIONS'
