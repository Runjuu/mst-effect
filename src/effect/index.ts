export * from './effect'
export { action, ValidEffectActions, EffectAction } from './action'
