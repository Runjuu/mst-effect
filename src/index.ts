export * from './types'
export * from './effect'
export * from './mst'
