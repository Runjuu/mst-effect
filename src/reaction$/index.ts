import { reaction, IReactionPublic, IReactionOptions } from 'mobx'
import { Observable } from 'rxjs'

export function reaction$<T>(
  expression: (r: IReactionPublic) => T,
  opts: IReactionOptions<T, true>,
): Observable<{ current: T; prev: T | undefined; r: IReactionPublic }>

export function reaction$<T>(
  expression: (r: IReactionPublic) => T,
  opts?: IReactionOptions<T, false>,
): Observable<{ current: T; prev: T; r: IReactionPublic }>

export function reaction$<T, FireImmediately extends boolean = false>(
  expression: (r: IReactionPublic) => T,
  opts?: IReactionOptions<T, FireImmediately>,
): Observable<{ current: T; prev?: T; r: IReactionPublic }> {
  return new Observable((observer) => {
    const dispose = reaction(
      expression,
      (current, prev, r) => observer.next({ current, prev, r }),
      opts,
    )

    return () => dispose()
  })
}
