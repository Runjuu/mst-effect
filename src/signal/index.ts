import { addDisposer } from 'mobx-state-tree'
import { Observable, Subject } from 'rxjs'

import { AnyInstance, PayloadFunc } from '../types'

export type SignalFactory<P, R> = (payload$: Observable<P>) => Observable<R>

export type SignalResult<P, R> = [Observable<R>, PayloadFunc<P, void>]

export function signal<P>(self: AnyInstance): SignalResult<P, P>
export function signal<P, R>(self: AnyInstance, fn: SignalFactory<P, R>): SignalResult<P, R>
export function signal<P, R>(self: AnyInstance, fn?: SignalFactory<P, R>): SignalResult<P, R> {
  const payloadSource = new Subject<P>()
  const signal$ = fn?.(payloadSource.asObservable()) ?? payloadSource.asObservable()

  addDisposer(self, () => payloadSource.complete())

  return [signal$, (payload: P) => payloadSource.next(payload)] as SignalResult<P, R>
}
