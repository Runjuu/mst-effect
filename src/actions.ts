import { EffectAction, ValidEffectActions } from './types'
import { Identity } from './const'

export function action<P extends any[]>(fn: (...params: P) => void, ...params: P): EffectAction {
  return { [Identity]: () => fn(...params) }
}

export function noop(..._params: any[]): EffectAction {
  return action(() => {})
}

export function runActions(actions: ValidEffectActions): void {
  if (Array.isArray(actions)) {
    actions.forEach(runAction)
  } else {
    runAction(actions)
  }
}

function isValidAction(action: any): action is EffectAction {
  return action && typeof action === 'object' && typeof action[Identity] === 'function'
}

function runAction(action: any): void {
  if (isValidAction(action)) {
    action[Identity]()
  } else if (action !== null) {
    console.warn(`[mst-effect]: ${action} is not a valid EffectActions`)
  }
}
