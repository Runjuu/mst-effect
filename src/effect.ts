import { Subject } from 'rxjs'
import { tap, catchError } from 'rxjs/operators'
import { IAnyModelType, Instance, addDisposer } from 'mobx-state-tree'

import type { PayloadFunc, EffectFactory } from './types'
import { HANDLE_MST_EFFECT_ACTIONS } from './const'

export function effect<P = void>(
  self: Instance<IAnyModelType>,
  fn: EffectFactory<P>,
): PayloadFunc<P, void>
export function effect(
  self: Instance<IAnyModelType>,
  fn: EffectFactory<any>,
): (payload: any) => void {
  const payloadSource = new Subject()
  const effect$ = fn(payloadSource.asObservable())
  const subscription = effect$
    .pipe(
      tap((actions) => {
        if (self[HANDLE_MST_EFFECT_ACTIONS]) {
          self[HANDLE_MST_EFFECT_ACTIONS](actions)
        } else {
          console.warn(
            `[mst-effect]: Make sure the 'types' is imported from 'mst-effect' instead of 'mobx-state-tree'`,
          )
        }
      }),
      logAngIgnoreError(fn),
    )
    .subscribe()

  addDisposer(self, () => {
    payloadSource.complete()
    subscription.unsubscribe()
  })

  return (payload) => {
    payloadSource.next(payload)
  }
}

function logAngIgnoreError(factory: unknown) {
  return catchError((err, caught) => {
    /* eslint-disable no-console */
    console.group('[mst-effect]: error')
    console.log(factory)
    console.error(err)
    console.groupEnd()
    /* eslint-enable no-console */
    return caught
  })
}
