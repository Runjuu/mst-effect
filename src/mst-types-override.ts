import { types as mstTypes } from 'mobx-state-tree'

import { HANDLE_MST_EFFECT_ACTIONS } from './const'
import { runActions } from './actions'

export const types: typeof mstTypes = {
  ...mstTypes,
  model(...params: any[]) {
    return mstTypes.model(...params).actions(() => ({
      [HANDLE_MST_EFFECT_ACTIONS]: runActions,
    }))
  },
}
