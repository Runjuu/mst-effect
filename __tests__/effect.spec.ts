import { from, timer, Subject, Observable } from 'rxjs'
import { map, startWith, endWith, switchMap, debounceTime, scan, filter } from 'rxjs/operators'
import { types as mstTypes } from 'mobx-state-tree'

import { types, effect, action, destroy, NOOP, ValidEffectActions } from '../src'

describe('effect', () => {
  describe(`action`, () => {
    it(`should execute the action immediately`, () => {
      const spy = jest.fn()
      const Model = types.model().actions((self) => ({
        log: effect<string>(self, (payload$) =>
          payload$.pipe(map((message) => action(spy, message))),
        ),
      }))

      const model = Model.create()

      model.log('hi!')

      expect(spy.mock.calls).toEqual([['hi!']])
    })

    it(`should accept 'NOOP' as valid action`, () => {
      const spy = jest.fn()
      const Model = types.model().actions((self) => ({
        run: effect<ValidEffectActions>(self, (payload$) => payload$),
      }))

      const model = Model.create()

      model.run(NOOP)
      model.run(action(spy))
      expect(spy.mock.calls).toEqual([[]])
    })

    it(`should accept an array of actions`, () => {
      const spy = jest.fn()
      const Model = types.model().actions((self) => ({
        log: effect<number>(self, (payload$) =>
          payload$.pipe(
            map((times) => Array.from({ length: times }).map((_, index) => action(spy, index))),
          ),
        ),
      }))

      const model = Model.create()

      model.log(3)
      expect(spy.mock.calls).toEqual([[0], [1], [2]])
    })

    it(`should ignore 'NOOP' inside action array`, () => {
      const spy = jest.fn()
      const Model = types.model().actions((self) => ({
        log: effect<string>(self, (payload$) =>
          payload$.pipe(map((message) => [NOOP, action(spy, message), NOOP])),
        ),
      }))

      const model = Model.create()

      model.log('hi!')
      expect(spy.mock.calls).toEqual([['hi!']])
    })

    it(`should able to modify model inside action`, () => {
      jest.useFakeTimers()
      const Model = types.model({ value: types.string }).actions((self) => ({
        setValue: effect<string>(self, (payload$) => {
          function updateValue(value: string) {
            self.value = value
          }

          return payload$.pipe(
            switchMap((newValue) => timer(1000).pipe(map(() => action(updateValue, newValue)))),
          )
        }),
      }))

      const model = Model.create({ value: '' })

      model.setValue('123')

      jest.advanceTimersByTime(1000)

      expect(model.value).toBe('123')
      jest.useRealTimers()
    })

    it(`should warning if action invalid`, () => {
      const spy = jest.spyOn(console, 'warn').mockImplementation()
      const Model = types.model().actions((self) => ({
        emit: effect<any>(self, (payload$) => payload$),
      }))

      const model = Model.create()
      const invalidActions = [
        'string',
        123,
        true,
        false,
        undefined,
        null,
        Symbol('invalid action'),
        {},
        function invalidAction() {},
      ]

      invalidActions.forEach((invalidAction) => model.emit(invalidAction))
      expect(spy.mock.calls.length).toBe(invalidActions.length)
      expect(spy.mock.calls).toMatchSnapshot()
      spy.mockRestore()
    })

    it(`should warning if 'types' is not imported from 'mst-effect'`, () => {
      const spy = jest.spyOn(console, 'warn').mockImplementation()
      const Model = mstTypes.model().actions((self) => ({
        emit: effect(self, new Subject()),
      }))

      Model.create()
      expect(spy.mock.calls).toMatchSnapshot()
      spy.mockRestore()
    })
  })

  describe(`subscription`, () => {
    it(`should subscribe the Observable when create`, () => {
      const spy = jest.fn()
      const Model = types.model().actions((self) => ({
        log: effect<never>(self, (payload$) => payload$.pipe(startWith(action(spy, 'start')))),
      }))

      Model.create()

      expect(spy.mock.calls).toEqual([['start']])
    })

    it(`should complete the Observable when destroy`, () => {
      const spy = jest.fn()
      const Model = types.model().actions((self) => ({
        log: effect<never>(self, (payload$) => payload$.pipe(endWith(action(spy, 'end')))),
      }))

      const model = Model.create()
      destroy(model)
      expect(spy.mock.calls).toEqual([['end']])
    })

    it(`should unsubscribe the Observable when destroy`, () => {
      const spy = jest.fn()
      const Model = types.model().actions((self) => ({
        log: effect<never>(self, () => new Observable(() => spy)),
      }))

      const model = Model.create()
      destroy(model)
      expect(spy).toBeCalledTimes(1)
    })
  })

  describe(`dispatch function return value`, () => {
    it(`should be a Promise`, () => {
      const Model = types.model().actions((self) => ({
        sendMsg: effect<string, string>(self, (payload$, resolve) =>
          payload$.pipe(map((message) => action(resolve, message))),
        ),
      }))

      const model = Model.create()
      expect(model.sendMsg('hi!')).toBeInstanceOf(Promise)
    })

    it(`should be resolved when trigger 'resolve'`, async () => {
      const Model = types.model().actions((self) => ({
        sendMsg: effect<string, string>(self, (payload$, resolve) =>
          payload$.pipe(map((message) => action(resolve, `message: ${message}`))),
        ),
      }))

      const model = Model.create()
      const msg = await model.sendMsg('hi!')
      expect(msg).toBe(`message: hi!`)
    })

    it(`should isolate resolved value when dispatch multi-times`, async () => {
      const Model = types.model().actions((self) => ({
        sendMsg: effect<string, string>(self, (payload$, resolve) =>
          payload$.pipe(map((message) => action(resolve, message))),
        ),
      }))

      const model = Model.create()

      const a = model.sendMsg('a')
      const b = model.sendMsg('b')
      const c = model.sendMsg('c')
      const result = await Promise.all([a, b, c])
      expect(result).toEqual(['a', 'b', 'c'])
    })

    it(`should resolve all previous result when trigger 'resolve'`, async () => {
      const Model = types.model().actions((self) => ({
        sendMsg: effect<string, string>(self, (payload$, resolve) =>
          payload$.pipe(
            debounceTime(10),
            map((message) => action(resolve, message)),
          ),
        ),
      }))

      const model = Model.create()

      const a = model.sendMsg('a')
      const b = model.sendMsg('b')
      const c = model.sendMsg('c')
      const result = await Promise.all([a, b, c])
      expect(result).toEqual(['c', 'c', 'c'])
    })

    it(`should resolve with undefined if did not trigger 'resolve' but the model was destroyed`, async () => {
      const Model = types.model().actions((self) => ({
        sendMsg: effect<string>(self, (payload$) => payload$.pipe(map(() => NOOP))),
      }))

      const model = Model.create()
      const promise = model.sendMsg('a')

      destroy(model)
      expect(await promise).toBeUndefined()
    })

    it(`should able to use RxJS to process the resolve value`, async () => {
      const Model = types.model().actions((self) => ({
        dispatch: effect<void, number>(self, (payload$, resolve) =>
          payload$.pipe(
            switchMap(() => from([1, 2, 3, 4]).pipe(map((num) => action(resolve, num)))),
          ),
        ),
      }))

      const model = Model.create()
      const result = await model.dispatch(undefined, (result$) =>
        result$.pipe(
          scan((result, value) => [...result, `${value}`], Array.of<string>()),
          filter((result) => result.length === 4),
        ),
      )

      expect(result).toEqual(['1', '2', '3', '4'])
    })

    it(`should able to use 'endWith' to provide default value`, async () => {
      const Model = types.model().actions((self) => ({
        sendMsg: effect<void, string>(self, (payload$, resolve) =>
          payload$.pipe(
            map(() => NOOP),
            endWith(action(resolve, 'end value')),
          ),
        ),
      }))

      const model = Model.create()
      const promise = model.sendMsg()

      destroy(model)
      expect(await promise).toBe('end value')
    })
  })

  it(`should keep alive after error occur`, () => {
    const spy = jest.fn()
    const spyError = jest.spyOn(console, 'error').mockImplementation()

    const Model = types.model().actions((self) => ({
      throwErr: effect<boolean>(self, (payload$) =>
        payload$.pipe(
          map((throwErr) => {
            if (throwErr) {
              throw new Error('')
            } else {
              return action(spy)
            }
          }),
        ),
      ),
    }))

    const model = Model.create()

    model.throwErr(true)
    expect(spy.mock.calls).toEqual([])
    expect(spyError.mock.calls.length).toBe(1)

    model.throwErr(false)
    expect(spy.mock.calls).toEqual([[]])
  })

  it(`should able to use Observable directly`, () => {
    const subject = new Subject<ValidEffectActions>()
    const spy = jest.fn()
    const Model = types
      .model({
        value: types.string,
      })
      .actions((self) => {
        effect(self, subject)
        return {}
      })

    Model.create({ value: '' })

    subject.next(action(spy, 123))

    expect(spy.mock.calls).toEqual([[123]])
  })
})
