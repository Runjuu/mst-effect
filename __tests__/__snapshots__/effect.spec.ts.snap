// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`effect action should warning if 'types' is not imported from 'mst-effect' 1`] = `
Array [
  Array [
    "[mst-effect]: Make sure the 'types' is imported from 'mst-effect' instead of 'mobx-state-tree'",
  ],
]
`;

exports[`effect action should warning if action invalid 1`] = `
Array [
  Array [
    "[mst-effect]: hi! is not a valid EffectActions",
  ],
]
`;
